Userland/Libraries/LibWeb/HTML/WorkerAgent.cpp

```diff
@@ -0,0 +1,114 @@
+}
+
+ErrorOr<Vector<String>> get_paths_for_helper_process(StringView process_name)
+{
+    auto application_path = TRY(application_directory());
+    Vector<String> paths;
+
+    TRY(paths.try_append(TRY(String::formatted("{}/{}", application_path, process_name))));
+    TRY(paths.try_append(TRY(String::formatted("./{}", process_name))));
+    // NOTE: Add platform-specific paths here
+    return paths;
+}
+
+ErrorOr<NonnullRefPtr<WebView::WebContentClient>> launch_web_content_process(ReadonlySpan<String> candidate_web_content_paths)
+{
+    int socket_fds[2] {};
+    TRY(Core::System::socketpair(AF_LOCAL, SOCK_STREAM, 0, socket_fds));
+
+    int ui_fd = socket_fds[0];
+    int wc_fd = socket_fds[1];
+
+    int fd_passing_socket_fds[2] {};
+    TRY(Core::System::socketpair(AF_LOCAL, SOCK_STREAM, 0, fd_passing_socket_fds));
+
+    int ui_fd_passing_fd = fd_passing_socket_fds[0];
+    int wc_fd_passing_fd = fd_passing_socket_fds[1];
+
+    if (auto child_pid = TRY(Core::System::fork()); child_pid == 0) {
+        TRY(Core::System::close(ui_fd_passing_fd));
+        TRY(Core::System::close(ui_fd));
+
+        auto takeover_string = TRY(String::formatted("WebContent:{}", wc_fd));
+        TRY(Core::System::setenv("SOCKET_TAKEOVER"sv, takeover_string, true));
+
+        auto webcontent_fd_passing_socket_string = TRY(String::number(wc_fd_passing_fd));
+
+        ErrorOr<void> result;
+        for (auto const& path : candidate_web_content_paths) {
+            if (Core::System::access(path, X_OK).is_error())
+                continue;
+
+            auto arguments = Vector {
+                path.bytes_as_string_view(),
+                "--webcontent-fd-passing-socket"sv,
+                webcontent_fd_passing_socket_string
+            };
+
+            result = Core::System::exec(arguments[0], arguments.span(), Core::System::SearchInPath::Yes);
+            if (!result.is_error())
+                break;
+        }
+
+        if (result.is_error())
+            warnln("Could not launch any of {}: {}", candidate_web_content_paths, result.error());
+        VERIFY_NOT_REACHED();
+    }
+
+    TRY(Core::System::close(wc_fd_passing_fd));
+    TRY(Core::System::close(wc_fd));
+
+    auto socket = TRY(Core::LocalSocket::adopt_fd(ui_fd));
+    TRY(socket->set_blocking(true));
+
+    auto new_client = TRY(adopt_nonnull_ref_or_enomem(new (nothrow) WebView::WebContentClient(move(socket))));
+    new_client->set_fd_passing_socket(TRY(Core::LocalSocket::adopt_fd(ui_fd_passing_fd)));
+
+    return new_client;
+}
+}
+#endif
+
+namespace Web::HTML {
+
+WorkerAgent::WorkerAgent(AK::URL url, WorkerOptions const& options)
+    : m_worker_options(options)
+    , m_url(move(url))
+{
+#ifndef AK_OS_SERENITY
+    // FIXME: Add factory function
+    auto paths = MUST(get_paths_for_helper_process("WebContent"sv));
+    m_worker_ipc = MUST(launch_web_content_process(paths));
+#else
+    m_worker_ipc = MUST(WebView::WebContentClient::try_create());
```

Comment:
From a quick look at the build failure - LibWeb cannot depend on LibWebView

Reply:
Bleh. I was hoping we could hack around that. Looks like this'll need a custom WebContentClient then

Reply:
Could this instead be an IPC that the chrome layer receives to create a worker process?